In [60]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [2]:
df = pd.read_csv('heart.csv')

In [3]:
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [4]:
df.shape

(918, 12)

In [5]:
df.isna().sum()

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

# so we are applying z-score

## for 3 column RestingBP,Cholesterol,MaxHR

In [25]:
df[df.RestingBP>(df.RestingBP.mean()+3*df.RestingBP.std())] 

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
109   39   M           ATA        190          241          0     Normal   
241   54   M           ASY        200          198          0     Normal   
365   64   F           ASY        200            0          0     Normal   
399   61   M           NAP        200            0          1         ST   
592   61   M           ASY        190          287          1        LVH   
732   56   F           ASY        200          288          1        LVH   
759   54   M           ATA        192          283          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
109    106              N      0.0       Up             0  
241    142              Y      2.0     Flat             1  
365    140              Y      1.0     Flat             1  
399     70              N      0.0     Flat             1  
592    150              Y      2.0     Down             1  
732    133              Y      4.0     Down             1  
759    195              N      0.0       Up             1

In [28]:
df[df.RestingBP<(df.RestingBP.mean()-3*df.RestingBP.std())] 

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
449   55   M           NAP          0            0          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
449    155              N      1.5     Flat             1

In [29]:
df1 = df[df.RestingBP<=(df.RestingBP.mean() +3* df.RestingBP.std())]
df1.shape

(911, 12)

In [30]:
df2 = df1[df1.RestingBP>=(df1.RestingBP.mean() -3 * df1.RestingBP.std())]
df2.shape

(910, 12)

In [31]:
df2[df2.Cholesterol>(df2.Cholesterol.mean() +3* df2.Cholesterol.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
76    32   M           ASY        118          529          0     Normal   
149   54   M           ASY        130          603          1     Normal   
616   67   F           NAP        115          564          0        LVH   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
76     130              N      0.0     Flat             1  
149    125              Y      1.0     Flat             1  
616    160              N      1.6     Flat             0

In [34]:
df2[df2.Cholesterol<(df2.Cholesterol.mean() -3* df2.Cholesterol.std())]

Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

### no outliers in negative

In [36]:
df3 = df2[df2.Cholesterol<=(df2.Cholesterol.mean() +3 * df.Cholesterol.std())]
df3.shape

(907, 12)

In [39]:
df3[df3.MaxHR >(df3.MaxHR.mean() +3 * df3.MaxHR.std())]


Empty DataFrame
Columns: [Age, Sex, ChestPainType, RestingBP, Cholesterol, FastingBS, RestingECG, MaxHR, ExerciseAngina, Oldpeak, ST_Slope, HeartDisease]
Index: []

In [40]:
df3[df3.MaxHR <(df3.MaxHR.mean() -3 * df3.MaxHR.std())]

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  \
390   51   M           ASY        140            0          0     Normal   

     MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
390     60              N      0.0     Flat             1

In [42]:
df4 = df3[df3.MaxHR >= (df3.MaxHR.mean() -3 * df3.MaxHR.std())]
df4.shape

(906, 12)

## df4 is our final

In [43]:
df4.shape

(906, 12)

In [44]:
df.shape

(918, 12)

In [45]:
df4.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [50]:
df4_w_d = pd.get_dummies(df4,drop_first=True,dtype=int)


In [51]:
df4_w_d.astype(int)

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease  \
0     40        140          289          0    172        0             0   
1     49        160          180          0    156        1             1   
2     37        130          283          0     98        0             0   
3     48        138          214          0    108        1             1   
4     54        150          195          0    122        0             0   
..   ...        ...          ...        ...    ...      ...           ...   
913   45        110          264          0    132        1             1   
914   68        144          193          1    141        3             1   
915   57        130          131          0    115        1             1   
916   57        130          236          0    174        0             1   
917   38        138          175          0    173        0             0   

     Sex_M  ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA  \
0        1                  1                  0                 0   
1        0                  0                  1                 0   
2        1                  1                  0                 0   
3        0                  0                  0                 0   
4        1                  0                  1                 0   
..     ...                ...                ...               ...   
913      1                  0                  0                 1   
914      1                  0                  0                 0   
915      1                  0                  0                 0   
916      0                  1                  0                 0   
917      1                  0                  1                 0   

     RestingECG_Normal  RestingECG_ST  ExerciseAngina_Y  ST_Slope_Flat  \
0                    1              0                 0              0   
1                    1              0                 0              1   
2                    0              1                 0              0   
3                    1              0                 1              1   
4                    1              0                 0              0   
..                 ...            ...               ...            ...   
913                  1              0                 0              1   
914                  1              0                 0              1   
915                  1              0                 1              1   
916                  0              0                 0              1   
917                  1              0                 0              0   

     ST_Slope_Up  
0              1  
1              0  
2              1  
3              0  
4              1  
..           ...  
913            0  
914            0  
915            0  
916            0  
917            1  

[906 rows x 16 columns]

In [52]:
df4_w_d.head()

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease  \
0   40        140          289          0    172      0.0             0   
1   49        160          180          0    156      1.0             1   
2   37        130          283          0     98      0.0             0   
3   48        138          214          0    108      1.5             1   
4   54        150          195          0    122      0.0             0   

   Sex_M  ChestPainType_ATA  ChestPainType_NAP  ChestPainType_TA  \
0      1                  1                  0                 0   
1      0                  0                  1                 0   
2      1                  1                  0                 0   
3      0                  0                  0                 0   
4      1                  0                  1                 0   

   RestingECG_Normal  RestingECG_ST  ExerciseAngina_Y  ST_Slope_Flat  \
0                  1              0                 0              0   
1                  1              0                 0              1   
2                  0              1                 0              0   
3                  1              0                 1              1   
4                  1              0                 0              0   

   ST_Slope_Up  
0            1  
1            0  
2            1  
3            0  
4            1

In [53]:
df4_w_d.shape

(906, 16)

In [54]:
X = df4_w_d.drop('HeartDisease',axis=1)

In [55]:
y= df4_w_d['HeartDisease']

In [56]:
X.shape

(906, 15)

In [57]:
y.shape

(906,)

## so we are done know use 1st Logistic Regression

In [58]:
from sklearn.linear_model import LogisticRegression

In [65]:
Logistic = LogisticRegression(max_iter=2000)

In [66]:
X_train , X_test ,y_train , y_test = train_test_split(X,y,test_size=0.2)

In [67]:
Logistic.fit(X_train,y_train)

LogisticRegression(max_iter=2000)

In [68]:
Logistic.score(X_test,y_test)

0.8296703296703297

# so using the Logistic we get this score 0.8296703296703297

## um lets check the result after the scaling

In [69]:
from sklearn.preprocessing import StandardScaler

In [70]:
scaler = StandardScaler()

In [71]:
X_scaled = scaler.fit_transform(X)

In [72]:
X_scaled

array([[-1.43268734,  0.46074981,  0.848846  , ..., -0.8232726 ,
        -0.99779493,  1.14251631],
       [-0.47787471,  1.62434375, -0.16934854, ..., -0.8232726 ,
         1.00220995, -0.87526103],
       [-1.75095822, -0.12104716,  0.79279859, ..., -0.8232726 ,
        -0.99779493,  1.14251631],
       ...,
       [ 0.37084763, -0.12104716, -0.62706901, ...,  1.2146645 ,
         1.00220995, -0.87526103],
       [ 0.37084763, -0.12104716,  0.35376058, ..., -0.8232726 ,
         1.00220995, -0.87526103],
       [-1.64486793,  0.34439041, -0.21605471, ..., -0.8232726 ,
        -0.99779493,  1.14251631]])

In [73]:
X_train , X_test ,y_train , y_test = train_test_split(X_scaled,y,test_size=0.2)

In [74]:
Logistic.fit(X_train,y_train)

LogisticRegression(max_iter=2000)

In [76]:
Logistic.score(X_test,y_test)

0.8626373626373627

# so little bit increases the model accuracy 0.8626373626373627

## if want to increase the modle accuracy more then try ::
### gridcv to find best HyperParameters for Logistic Regression model & also we can try differnt algorithms also

In [79]:
from sklearn.ensemble import RandomForestClassifier

In [80]:
Random = RandomForestClassifier()

In [81]:
Random.fit(X_train,y_train)

RandomForestClassifier()

In [82]:
Random.score(X_test,y_test)

0.8351648351648352

## ok this also good if we do little for search for best HyperParameters then this model also give good results

# So lets Do the PCA

In [84]:
from sklearn.decomposition import PCA

In [120]:
pca = PCA(0.95)

In [121]:
pca_X = pca.fit_transform(X)

In [122]:
pca_X.shape

(906, 2)

In [123]:
X_train , X_test ,y_train,y_test = train_test_split(pca_X,y,test_size=0.2)

In [124]:
Logistic.fit(X_train,y_train)

LogisticRegression(max_iter=2000)

In [125]:
Logistic.score(X_test,y_test)

0.7307692307692307

In [126]:
Random.fit(X_train,y_train)

RandomForestClassifier()

In [127]:
Random.score(X_test,y_test)

0.7197802197802198

## um so based uppon the PCA result we can say that all columns (almost) are important to predict the result 